compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseConsumerIntoLoop.cpp

```diff
@@ -83,22 +97,46 @@ void AMDAIEFuseConsumerIntoLoopPass::runOnOperation() {
     return;
   }
 
-  // Step 4. Greedily fuse the consumer ops for a specified fusion depth and
-  // while fusion keeps occurring or until the maximum number of iterations is
-  // exceeded.
-  bool changed{true};
-  int64_t iter{0};
-  while (changed && iter < maxIterations) {
-    changed = false;
-    // Canonicalize before every iteration to enable more back-to-back fusion
-    // opportunities.
-    (void)applyPatternsGreedily(funcOp, canonicalizationPatterns);
-    Operation *producerOp = computeOp;
+  // Step 4. Find outermost nested scf loop and maintain the loop nest count.
+  int64_t loopNestLength = 1;
+  do {
+    ResultRange results = loop->getResults();
+    SmallVector<Operation *> allUsers = std::accumulate(
+        results.begin(), results.end(), SmallVector<Operation *>{},
+        [](SmallVector<Operation *> init, OpResult res) {
+          for (Operation *op : res.getUsers()) init.push_back(op);
+          return init;
+        });
+    if (allUsers.size() != 1) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Expected only one user of the compute/loop op.\n");
+      return;
+    }
+    if (!(isa<tensor::InsertSliceOp, tensor::ParallelInsertSliceOp>(
+            allUsers[0]))) {
+      break;
+    }
+    computeOp = loop;
+    loop = loop->getParentOfType<LoopLikeOpInterface>();
+    loopNestLength++;
+  } while (true);
+
+  // Step 5. Greedily fuse the consumer ops for a specified fusion depth at each
+  // level of the loopnest.
+  bool changedGlobal{false};
+  bool changedLocal{true};
+  while (loopNestLength != 0) {
+    changedLocal = false;
+    if (loopNestLength > 1) {
+      computeOp = findPostOrderComputeOpInLoop<LoopLikeOpInterface>(loop);
+    } else {
+      computeOp = findPostOrderComputeOpInLoop<linalg::LinalgOp>(loop);
+    }
```

Comment:
I think we should check that `computeOp` is not a nullptr at this point.

Reply:
Done.